### PR TITLE
Unity Adapter Update for UnityAds SDK v3.5.0

### PR DESCRIPTION
**Pull request by AliceNoWonderland**
Adapter updates for UnityAdsSDK v3.5.0.

Add IUnityAdsInitializationListener get UnityAds Initialization callbacks.
Add IUnityAdsLoadListener to get UnityAds load() callbacks.
Limited the usage of original IUnityAdsExtendedListener for UnityAds show() callbacks only.
Remove UnityIntersitialCallbackRouter class

### CODE REVIEW

File: UnityAds/src/main/java/com/mopub/mobileads/UnityInterstitial.java [hunk 1]
```diff
@@ -136,55 +146,34 @@ public void onUnityAdsFinish(String placementId, UnityAds.FinishState finishStat
...
-            MoPubLog.log(CUSTOM, ADAPTER_NAME, "Unity interstitial video cache failed for placement " +
-                    mPlacementId + "." + message);
-            MoPubErrorCode errorCode = UnityRouter.UnityAdsUtils.getMoPubErrorCode(unityAdsError);
-            mLoadListener.onAdLoadFailed(errorCode);
-
-            MoPubLog.log(LOAD_FAILED, ADAPTER_NAME,
-                    errorCode.getIntCode(),
-                    errorCode);
+        if (unityAdsError == SHOW_ERROR) {
+            UnityAds.removeListener(UnityInterstitial.this);
```
Comment (hmertkahraman) [inline UnityAds/src/main/java/com/mopub/mobileads/UnityInterstitial.java:176]: 1- When is this method getting called currently? Is it when an interstitial is attempted to be shown but failed to show for a reason (explained with `String message`?

If so, do you also think we should also include the fail to show logic inside `show()` method which is:

```
// Lets Unity Ads know when ads fail to show
            MediationMetaData metadata = new MediationMetaData(mContext);
            metadata.setMissedImpressionOrdinal(++missedImpressionOrdinal);
            metadata.commit();

            MoPubLog.log(CUSTOM, ...

            MoPubLog.log(SHOW_FAILED, ADAPTER_NAME,
                    ...
```

2- What other errors can be raised on this callback? Should we expect to cover some other error messages?

File: UnityAds/src/main/java/com/mopub/mobileads/UnityAdsAdapterConfiguration.java [hunk 2]
```diff
@@ -62,32 +62,37 @@ public void initializeNetwork(@NonNull final Context context, @Nullable final Ma
...
-
-                    networkInitializationSucceeded = true;
-                } else {
+                    listener.onNetworkInitializationFinished(UnityAdsAdapterConfiguration.class, MoPubErrorCode.ADAPTER_INITIALIZATION_SUCCESS);
+                    return;
+                }
+                if (configuration == null) {
                     MoPubLog.log(CUSTOM, ADAPTER_NAME, "Unity Ads initialization not started. " +
-                            "Context is not an Activity. Note that initialization on the first app launch is a no-op.");
+                            "Configuration is null. Note that initialization on the first app launch is a no-op.");
```
[the remark this answers is not in the document]
Comment (AliceNoWonderland) [reply UnityAds/src/main/java/com/mopub/mobileads/UnityAdsAdapterConfiguration.java:73]: Yes, that is correct @hmertkahraman

File: UnityAds/src/main/java/com/mopub/mobileads/UnityAdsAdapterConfiguration.java [hunk 3]
```diff
@@ -62,32 +62,37 @@ public void initializeNetwork(@NonNull final Context context, @Nullable final Ma
...
-        if (networkInitializationSucceeded) {
-            listener.onNetworkInitializationFinished(UnityAdsAdapterConfiguration.class,
-                    MoPubErrorCode.ADAPTER_INITIALIZATION_SUCCESS);
-        } else {
-            listener.onNetworkInitializationFinished(UnityAdsAdapterConfiguration.class,
-                    MoPubErrorCode.ADAPTER_CONFIGURATION_ERROR);
-        }
 
         MoPubLog.LogLevel logLevel = MoPubLog.getLogLevel();
         boolean debugModeEnabled = logLevel == MoPubLog.LogLevel.DEBUG;
```
[the remark this answers is not in the document]
Comment (AliceNoWonderland) [reply UnityAds/src/main/java/com/mopub/mobileads/UnityAdsAdapterConfiguration.java:98]: Yes, it can be safely called without SDK initialized

**Comment by hmertkahraman** [comment]
Also what do you think of the default placement Ids assigned to the adapters?
Looking at three adapters we can see:
- Banner:
```
private String mPlacementId = "";
```
- Interstitial:
```
private String mPlacementId = "";
```

- Rewarded Video:
```
private String mPlacementId = "";
```

**Comment by hmertkahraman** [comment]
Another topic is that RV adapter is executing the initialization logic inside a 
```
synchronized (UnityRewardedVideo.class) {
```
block, while interstitial and banner does not. Do you think there's an advantage or a need of having the `synchronized` statement, and should we ensure parity across them all, or is it needed only on RV?

**Comment by hmertkahraman** [comment]
Since initialization method now can be called with `context` instead of `activity`, can we safely remove the `if activity` check on Banner adapters [here](https://github.com/mopub/mopub-android-mediation/pull/300/files#diff-2c2424d4957238e40fcb29dcfce75e4919d9513eaa1d8dc6cf3ce871ba56d786R46-R55)?

File: UnityAds/src/main/java/com/mopub/mobileads/UnityInterstitial.java (hunk 1 above)
Comment (AliceNoWonderland) [reply UnityAds/src/main/java/com/mopub/mobileads/UnityInterstitial.java:176]: @hmertkahraman `onUnityAdsError` is triggered on a lot of different unit ads errors but majority is related to initialization and load, so only leave show error for this to track. Not sure if I understand 

> If so, do you also think we should also include the fail to show logic inside show() method which is:

This implementation should cover all unity errors, load errors are tracked by load callbacks, init errors are tracked by init callbacks and show errors are tracked by `onUnityAdsFinish`/`onUnityAdsError `

**Comment by AliceNoWonderland** [comment]
> Also what do you think of the default placement Ids assigned to the adapters?
> Looking at three adapters we can see:
> 
> * Banner (looks fine):
> 
> ```
> private String placementId = "banner";
> ```
> 
> * Interstitial (should this remain as `video`?):
> 
> ```
> private String mPlacementId = "video";
> ```
> 
> * Rewarded Video (should we change this to `video` as well?):
> 
> ```
> private String mPlacementId = "";
> ```

Yes, I can assign a default placementId to rewarded video (`rewardedVideo`), and the rest default placementId looks fine

**Comment by AliceNoWonderland** [comment]
> synchronized (UnityRewardedVideo.class) {

Yes, we can safely remove the `synchronized` block now 👍

**Comment by AliceNoWonderland** [comment]
> Since initialization method now can be called with `context` instead of `activity`, can we safely remove the `if activity` check on Banner adapters [here](https://github.com/mopub/mopub-android-mediation/pull/300/files#diff-2c2424d4957238e40fcb29dcfce75e4919d9513eaa1d8dc6cf3ce871ba56d786R46-R55)?

Actually, even though our initialization method accept context, the`BannerView` still only accept activity context, so the if check here is still valid

**Comment by hmertkahraman** [comment]
Released internally [here](https://github.com/mopub/mopub-android-mediation/releases/tag/unityads-3.5.0.0) with improvements on initialization and overall adapter functions

**Pull request closed without merging**